File: sdkwork-rtc-sdk-go/rtcstandard/provider_selection.go
package rtcstandard

import "strings"

type ParsedRtcProviderUrl struct {
    ProviderKey string
    RawUrl      string
}

type RtcProviderSelection struct {
    ProviderKey string
    Source      string
}

type RtcProviderSelectionRequest struct {
    ProviderUrl                  string
    ProviderKey                  string
    TenantOverrideProviderKey    string
    DeploymentProfileProviderKey string
}

var RtcProviderSelectionSources = []string{
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
}

var RtcProviderSelectionPrecedence = []string{
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
}

func hasRtcProviderSelectionText(value string) bool {
    return strings.TrimSpace(value) != ""
}

func ParseRtcProviderUrl(providerUrl string) ParsedRtcProviderUrl {
    trimmed := strings.TrimSpace(providerUrl)
    if !strings.HasPrefix(trimmed, "rtc:") || !strings.Contains(trimmed, "://") {
        panic("Invalid RTC provider URL: " + providerUrl)
    }

    withoutPrefix := strings.TrimPrefix(trimmed, "rtc:")
    providerKey, _, _ := strings.Cut(withoutPrefix, "://")

    return ParsedRtcProviderUrl{
        ProviderKey: strings.ToLower(providerKey),
        RawUrl:      providerUrl,
    }
}

func ResolveRtcProviderSelection(
    request RtcProviderSelectionRequest,
    defaultProviderKey string,
) RtcProviderSelection {
    if hasRtcProviderSelectionText(request.ProviderUrl) {
        return RtcProviderSelection{
            ProviderKey: ParseRtcProviderUrl(request.ProviderUrl).ProviderKey,
            Source:      "provider_url",
        }
    }

    if hasRtcProviderSelectionText(request.ProviderKey) {
        return RtcProviderSelection{
            ProviderKey: strings.TrimSpace(request.ProviderKey),
            Source:      "provider_key",
        }
    }

    if hasRtcProviderSelectionText(request.TenantOverrideProviderKey) {
        return RtcProviderSelection{
            ProviderKey: strings.TrimSpace(request.TenantOverrideProviderKey),
            Source:      "tenant_override",
        }
    }

    if hasRtcProviderSelectionText(request.DeploymentProfileProviderKey) {
        return RtcProviderSelection{
            ProviderKey: strings.TrimSpace(request.DeploymentProfileProviderKey),
            Source:      "deployment_profile",
        }
    }

    resolvedDefaultProviderKey := defaultProviderKey
    if !hasRtcProviderSelectionText(resolvedDefaultProviderKey) {
        resolvedDefaultProviderKey = DEFAULT_RTC_PROVIDER_KEY
    }

    return RtcProviderSelection{
        ProviderKey: resolvedDefaultProviderKey,
        Source:      "default_provider",
    }
}

File: sdkwork-rtc-sdk-typescript/src/provider-selection.ts
import { RtcSdkException } from './errors.js';
import { DEFAULT_RTC_PROVIDER_KEY } from './provider-catalog.js';
import { freezeRtcRuntimeValue } from './runtime-freeze.js';
import type {
  RtcProviderSelection,
  RtcProviderSelectionRequest,
  RtcProviderSelectionSource,
} from './types.js';

const RTC_PROVIDER_URL_PATTERN = /^rtc:([a-z0-9-]+):\/\/.+$/i;

export interface ParsedRtcProviderUrl {
  providerKey: string;
  rawUrl: string;
}

export const RTC_PROVIDER_SELECTION_SOURCES = freezeRtcRuntimeValue([
  'provider_url',
  'provider_key',
  'tenant_override',
  'deployment_profile',
  'default_provider',
] as const);

export const RTC_PROVIDER_SELECTION_PRECEDENCE = freezeRtcRuntimeValue([
  ...RTC_PROVIDER_SELECTION_SOURCES,
]);

export function parseRtcProviderUrl(providerUrl: string): ParsedRtcProviderUrl {
  const match = RTC_PROVIDER_URL_PATTERN.exec(providerUrl.trim());
  if (!match) {
    throw new RtcSdkException({
      code: 'invalid_provider_url',
      message: `Invalid RTC provider URL: ${providerUrl}`,
      details: { providerUrl },
    });
  }

  return freezeRtcRuntimeValue({
    providerKey: match[1].toLowerCase(),
    rawUrl: providerUrl,
  });
}

function createRtcProviderSelection(
  providerKey: string,
  source: RtcProviderSelectionSource,
): RtcProviderSelection {
  return freezeRtcRuntimeValue({
    providerKey,
    source,
  });
}

export function resolveRtcProviderSelection(
  request: RtcProviderSelectionRequest = {},
  defaultProviderKey: string = DEFAULT_RTC_PROVIDER_KEY,
): RtcProviderSelection {
  if (request.providerUrl) {
    return createRtcProviderSelection(
      parseRtcProviderUrl(request.providerUrl).providerKey,
      'provider_url',
    );
  }

  if (request.providerKey) {
    return createRtcProviderSelection(request.providerKey, 'provider_key');
  }

  if (request.tenantOverrideProviderKey) {
    return createRtcProviderSelection(
      request.tenantOverrideProviderKey,
      'tenant_override',
    );
  }

  if (request.deploymentProfileProviderKey) {
    return createRtcProviderSelection(
      request.deploymentProfileProviderKey,
      'deployment_profile',
    );
  }

  return createRtcProviderSelection(defaultProviderKey, 'default_provider');
}
